src/einsteinpy/utils/dual.py

```diff
@@ -0,0 +1,222 @@
+
+        return DualNumber(self.a / other, self.b / other)
+
+    def __rtruediv__(self, other):
+        if isinstance(other, DualNumber):
+            if self.a == 0 and other.a == 0:
+                return DualNumber(other.b / self.b, 0.0)
+
+            return DualNumber(
+                other.a / self.a, (other.b * self.a - other.a * self.b) / (self.a ** 2)
+            )
+
+        return DualNumber(other, 0).__truediv__(self)
+
+    def __eq__(self, other):
+        return (self.a == other.a) and (self.b == other.b)
+
+    def __ne__(self, other):
+        return not (self == other)
+
+    def __neg__(self):
+        return DualNumber(-self.a, -self.b)
+
+    def __pow__(self, power):
+        return DualNumber(self.a ** power, self.b * power * self.a ** (power - 1))
+
+    def sin(self):
+        return DualNumber(np.sin(self.a), self.b * np.cos(self.a))
+
+    def cos(self):
+        return DualNumber(np.cos(self.a), -self.b * np.sin(self.a))
+
+    def tan(self):
+        return np.sin(self) / np.cos(self)
+
+    def log(self):
+        return DualNumber(np.log(self.a), self.b / self.a)
+
+    def exp(self):
+        return DualNumber(np.exp(self.a), self.b * np.exp(self.a))
+
+
+def _deriv(func, x):
+    """
+    Calculates first (partial) derivative of ``func`` at ``x``
+
+    Parameters
+    ----------
+    func : callable
+        Function to differentiate
+    x : float
+        Point, at which, the derivative will be evaluated
+
+    Returns
+    _______
+    float
+        First partial derivative of ``func`` at ``x``
+
+    """
+    funcdual = func(DualNumber(x, 1.0))
+
+    if isinstance(funcdual, DualNumber):
+        return funcdual.b
+
+    return 0.0
+
+
+def _diff_g(g, g_prms, coords, indices, wrt):
+    """
+    Computes derivative of metric elements
+
+    Parameters
+    ----------
+    g : callable
+        Metric (Contravariant) Function
+    coords : array_like
+        4-Position
+    indices : array_like
+        2-tuple, containing indices, indexing a metric
+        element, whose derivative will be calculated
+    wrt : int
+        coordsinate, with respect to which, the derivative
+        will be calculated
+        Takes values from ``[0, 1, 2, 3]``
+
+    Returns
+    -------
+    float
+        Value of derivative of metric element at ``coords``
+
+    Raises
+    ------
+    ValueError
+        If ``wrt`` is not in [1, 2, 3, 4]
+        or ``len(indices) != 2``
+
+    """
+    if wrt not in [0, 1, 2, 3]:
+        raise ValueError(f"wrt takes values from [0, 1, 2, 3]. Supplied value: {wrt}")
+
+    if len(indices) != 2:
+        raise ValueError("indices must be a 2-tuple containing indices for the metric.")
+
+    dual_coords = [
+        DualNumber(coords[0], 0.0),
+        DualNumber(coords[1], 0.0),
+        DualNumber(coords[2], 0.0),
+        DualNumber(coords[3], 0.0),
+    ]
+
+    dual_coords[wrt].b = 1.0
+
+    return _deriv(lambda q: g(dual_coords, *g_prms)[indices], coords[wrt])
```

Comment:
I don't understand. `q` is the argument, but it is never used in the body of the lambda. So, how will `coords[wrt]` be passed to the `func` in `_deriv`?

Reply:
Ah, that's because `dual_coords` contains a `DualNumber` version of everything `coords` has, with the same indices. So, to continue a discussion from above, I could just as well rewrite this line as:
```
return _deriv(lambda q: g(dual_coords, *g_prms)[indices], dual_coords[wrt])
```
and remove the typecast into `DualNumber` from [here](https://github.com/JeS24/einsteinpy/blob/50f2a7edfdb0c1c08c14050950a498ced075730d/src/einsteinpy/utils/dual.py#L132):
```
funcdual = func(x)
```
But this makes `_deriv` very non-general / restricted.